### PR TITLE
fix: Page order sequence for git import

### LINKED ISSUE #12944: [Bug]: When we do git import the application does not have the same page sequence as in original git application.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

When we do git import the application does not have the same page sequence as in original git application.
image

image

### Steps To Reproduce

1. Create new application
2. Add pages
3. Connect app to git
4. Enow import git connect app from git import
5. check the page index it will be different from original app

### Public Sample App

_No response_

### Version

release

### PR DESCRIPTION
**Pull request by AnaghHegde**
## Description

> Page order sequence gets messed up when we do git import. The reason is because of the parallel execution of task by stream API, which is used to process page from the application json file. We don't have control over this. So we have decided to add the page order to the metadata of json file which is exported. This list is then referred while importing application into the org. 

Fixes #12944 

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

- Locally 
- Junit Tests

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/domains/ApplicationJson.java [hunk 1]
```diff
@@ -35,6 +35,8 @@ public class ApplicationJson {
 
     List<NewPage> pageList;
 
+    List<String> pageOrder;
```
Comment (abhvsn) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/domains/ApplicationJson.java:38]: We should have a seperate list for unpublished and published pages as the sequence might be different

Comment (AnaghHegde) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/domains/ApplicationJson.java:38]: @abhvsn IIUC, it is not possible right? Say you have Page1 at top in edit mode and in view mode it is at bottom. Now while you are creating a new App, the same page will have two different order and we have to ignore one of these two. And IMO, edit mode takes precedence over view mode.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java [hunk 2]
```diff
@@ -244,9 +244,14 @@ public Mono<ApplicationJson> exportApplicationById(String applicationId, Seriali
 
                     return pageFlux
```
Comment (abhvsn) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java:245]: This again result in mismatched order can we instead take the reference from `pages` and `publishedPages` from the application object that will be easier to deal with. Also we are referencing the pages from application object only

Comment (AnaghHegde) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java:245]: It does not matter now as we are storing the page order as a metadata, which is used in import flow. So no matter how the pages are stored in json-file/git-repo, the order is maintained as we keep update the order in metadata during the export flow here - https://github.com/appsmithorg/appsmith/blob/4a704d415f7abfe484114ddf2fb2c696f6d13345/app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java#L254.

**Review by abhvsn — COMMENTED** [review]
Please take a look at the case mentioned in the comment also can we add the testcase with the mentioned scenario

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java [hunk 3]
```diff
@@ -253,6 +254,13 @@ public Mono<ApplicationJson> exportApplicationById(String applicationId, Seriali
                                     pageOrder.add(page.getUnpublishedPage().getName());
                                 }
                                 applicationJson.setPageOrder(pageOrder);
+
+                                Collections.sort(newPages, Comparator.comparing(newPage -> publishedPageOrderList.indexOf(newPage.getId())));
```
[the remark this answers is not in the document]
Comment (AnaghHegde) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java:258]: @abhvsn This is handled correctly during the import. But when we add the page order to json file during export, i have missed out on this case. Let me fix this.

File: app/server/appsmith-server/src/test/java/com/appsmith/server/solutions/ImportExportApplicationServiceTests.java [hunk 4]
```diff
@@ -2359,22 +2359,30 @@ public void exportAndImportApplication_withMultiplePages_PagesOrderIsMaintained(
...
-                    List<NewPage> pageList = applicationJson.getPageList();
-                    assertThat(pageList.get(0).getUnpublishedPage().getName()).isEqualTo("123");
-                    assertThat(pageList.get(1).getUnpublishedPage().getName()).isEqualTo("abc");
-                    assertThat(pageList.get(2).getUnpublishedPage().getName()).isEqualTo("Page1");
+                    List<String> pageList = applicationJson.getPageOrder();
+                    assertThat(pageList.get(0)).isEqualTo("123");
+                    assertThat(pageList.get(1)).isEqualTo("abc");
+                    assertThat(pageList.get(2)).isEqualTo("Page1");
+
+                    List<String> publishedPageList = applicationJson.getPageOrder();
```
Comment (abhvsn) [inline app/server/appsmith-server/src/test/java/com/appsmith/server/solutions/ImportExportApplicationServiceTests.java:2375]:
```suggestion
                    List<String> publishedPageList = applicationJson.publishedPageOrder();
```

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (abhvsn) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java:258]: Now getting the NPE during the export

**Review by abhvsn — CHANGES_REQUESTED** [review]
Now I'm getting the NPE for the for the scenario described https://github.com/appsmithorg/appsmith/pull/13126/files#r857259455 at 
`pageOrder.add(page.getPublishedPage().getName());`

**Review by abhvsn — APPROVED** [review]

**Pull request merged**